test/persistence-hooks.suite.js

```diff
@@ -1374,6 +1374,246 @@ module.exports = function(dataSource, should) {
+
+        existingInstance.replaceAttributes({ name: 'changed' }, function(err) {
+          if (err) return done(err);
+          observedContexts.should.eql(aTestModelCtx({
+            instance: {
+              id: existingInstance.id,
+              name: 'changed',
+              extra: undefined,              
+            },
+            isNewInstance: false
+          }));
+          done();
+        });
+      });
+
+      it('aborts when `before save` hook fails', function(done) {
+        TestModel.observe('before save', nextWithError(expectedError));
+
+        existingInstance.replaceAttributes({ name: 'replaced' }, function(err) {
+          [err].should.eql([expectedError]);
+          done();
+        });
+      });
+
+      it.skip('applies updates from `before save` hook', function(done) {
+        /* Should they be able to change `instance` in `before save`? I was completely on the wrong track? :(
+         * in line [1] data is over writting instance!
+         * Also in line [2] I was copying `data` to `inst`
+         * Could you please clarify if in `before save` for `replaceAttributes` they are allowed to change instance? 
+         * 
+         * [1]:https://github.com/strongloop/loopback-datasource-juggler/pull/788/files#diff-d717a01e40d8164976e4468af1bce663R2678
+         * [2]:https://github.com/strongloop/loopback-datasource-juggler/pull/788/files#diff-d717a01e40d8164976e4468af1bce663R2707
+        */
+
+        TestModel.observe('before save', function(ctx, next) {
+          ctx.instance.name = 'hooked';
+          next();
+        });
+
+        existingInstance.replaceAttributes({ name: 'updated' }, function(err) {
+          if (err) return done(err);
+          TestModel.findById(existingInstance.id, function(err, instance) {
+            if (err) return done(err);
+            should.exists(instance);
+            instance.toObject(true).should.eql({
+              id: existingInstance.id,
+              name: 'hooked',
+              extra: undefined
+            });
+            done();
+          });
+        });
+      });      
+
+      it('triggers `persist` hook', function(done) {
+        TestModel.observe('persist', pushContextAndNext());
+        existingInstance.replaceAttributes({ name: 'replacedName' }, function(err) {
+          if (err) return done(err);
+
+          observedContexts.should.eql(aTestModelCtx({
+            where: { id: existingInstance.id },
+            data: {
+              name: 'replacedName',
+              id: existingInstance.id
+            },
+            currentInstance: {
+              id: existingInstance.id,
+              name: 'replacedName',
+              extra: null
+            },
+            isNewInstance: false
+          }));
+
+          done();
+        });
+      });
+      
+      it('applies updates from `persist` hook', function(done) {
+        TestModel.observe('persist', pushContextAndNext(function(ctx){
+          ctx.data.extra = 'hook data';
```

Comment:
Since one of the main purpose of REPLACE is to allow people to delete properties, I think it will be better to change test cases to delete an existing property instead of adding a new one.

In order to get this PR finished soon, let's keep this task out of scope, but make sure to address it in the near future.